src/VisualStudio/Core/Def/Implementation/FindUsages/VisualStudioFindSymbolMonikerUsagesService.cs

```diff
@@ -0,0 +1,121 @@
+using System.Collections.Immutable;
+using System.Composition;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.LanguageServiceIndexFormat;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Newtonsoft.Json.Linq;
+using VS.IntelliNav.Contracts;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    [ExportWorkspaceService(typeof(IFindSymbolMonikerUsagesService), layer: ServiceLayer.Host), Shared]
+    internal partial class VisualStudioFindSymbolMonikerUsagesService : AbstractFindSymbolMonikerUsagesService
+    {
+        private readonly ICodeIndexProvider? _codeIndexProvider;
+
+        private readonly object _gate = new object();
+        private CancellationTokenSource? _lastNavigationCancellationSource;
+
+        [ImportingConstructor]
+        public VisualStudioFindSymbolMonikerUsagesService(
+            [Import(AllowDefault = true)] ICodeIndexProvider? codeIndexProvider)
+        {
+            _codeIndexProvider = codeIndexProvider;
+        }
+
+        public override async IAsyncEnumerable<ExternalReferenceItem> FindReferencesByMoniker(
+            DefinitionItem definition, ImmutableArray<SymbolMoniker> monikers,
+            IStreamingProgressTracker progress, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            if (_codeIndexProvider == null)
+                yield break;
+
+            var convertedMonikers = ConvertMonikers(monikers);
+            var currentPage = 0;
+            while (true)
+            {
+                var referenceItems = await FindReferencesByMonikerAsync(
+                    _codeIndexProvider, definition, convertedMonikers, progress, currentPage, cancellationToken).ConfigureAwait(false);
+
+                // If we got no items, we're done.
+                if (referenceItems.Length == 0)
+                    break;
+
+                foreach (var item in referenceItems)
+                    yield return item;
+
+                // Otherwise, we got some items.  Return them to our caller and attempt to retrieve
+                // another page.
+                currentPage++;
+            }
+        }
+
+        private async Task<ImmutableArray<ExternalReferenceItem>> FindReferencesByMonikerAsync(
+            ICodeIndexProvider codeIndexProvider, DefinitionItem definition, ImmutableArray<ISymbolMoniker> monikers,
+            IStreamingProgressTracker progress, int pageIndex, CancellationToken cancellationToken)
+        {
+            // Let the find-refs window know we have outstanding work
+            await using var _1 = await progress.AddSingleItemAsync().ConfigureAwait(false);
```

Comment:
We can't do multiple discards like this? :-(

Reply:
nope.

Reply:
@jcouv 😢